### PR TITLE
[Incremental] Next step towards cross-module incremental dependencies

### PR DESCRIPTION
**Pull request by davidungar**
Lots of restructuring; passes the lit tests.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

File: Sources/SwiftDriver/IncrementalCompilation/SourceFileDependencyGraph.swift [hunk 1]
```diff
@@ -250,13 +270,18 @@ extension SourceFileDependencyGraph {
...
+      try data.contents.withUnsafeBytes { buf in
+        // SAFETY: The bitcode reader does not mutate the data stream we give it.
+        // FIXME: Let's avoid this altogether and traffic in ByteString/[UInt8]
+        // if possible. There's no real reason to use `Data` in this API.
+        let baseAddr = UnsafeMutableRawPointer(mutating: buf.baseAddress!)
+        let data = Data(bytesNoCopy: baseAddr, count: buf.count, deallocator: .none)
+        try Bitcode.read(stream: data, using: &visitor)
+      }
+    }
+    catch ReadError.swiftModuleHasNoDependencies {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/SourceFileDependencyGraph.swift:283]: Shouldn't we catch every kind of error here? Like, if more than the signature were to be corrupted and the visitor were to detect it, we should just say the graph is busted and start over.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift [hunk 2]
```diff
@@ -105,84 +106,96 @@ extension ModuleDependencyGraph {
...
+    guard let sourceGraph = dependencySource.read(
+            in: fileSystem,
+            reporter: reporter)
+    else {
+      return nil
+    }
+    return integrate(sourceGraph: sourceGraph)
+  }
+
+  public func integrate(sourceGraph: SourceFileDependencyGraph) -> Set<Node>? {
```
Comment (CodaFi) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:140]: This is a fantastic cleanup

**Review by CodaFi — APPROVED** [review]
Just one question, but it's not enough to hold back integrating this PR.

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift (hunk 2 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraph.swift:140]: thanks!

File: Sources/SwiftDriver/IncrementalCompilation/SourceFileDependencyGraph.swift (hunk 1 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/SourceFileDependencyGraph.swift:283]: I don't think so. This covers the special case of getting an external-dep with a fingerprint, but in the meantime the vendor has been compiled without the flag. But maybe I don't follow; we can discuss offline.

**Comment by davidungar** [comment]
@swift-ci please test macos platform

**Comment by CodaFi** [comment]
⛵

**Pull request merged**